nft_ingester/src/consistency_calculator.rs

```diff
@@ -0,0 +1,833 @@
+            } else if epoch > last_slot_epoch && last_slot != 0 {
+                let _ = sender
+                    .send(ConsistencyCalcMsg::EpochChanged { new_epoch: epoch })
+                    .await;
+            }
+        }
+    }
+
+    /// Iterates over bubblegum changes, and for each of them check if the change from an epoch
+    /// previous to the current. If the change is from the previous epoch,
+    /// it sends a notification for the checksums calculator.
+    /// This method is called from the fork cleaner.
+    pub async fn watch_remove_forked_bubblegum_changes(&self, keys: &[BubblegumChangeKey]) {
+        let last_slot = storage_consistency::last_tracked_slot();
+        let last_slot_epoch = epoch_of_slot(last_slot);
+        if let Some(sender) = self.sender.as_ref() {
+            for key in keys {
+                if key.epoch < last_slot_epoch {
+                    let _ = sender
+                        .send(ConsistencyCalcMsg::BubblegumUpdated {
+                            tree: key.tree_pubkey,
+                            slot: key.slot,
+                        })
+                        .await;
+                }
+            }
+        }
+    }
+}
+
+/// An entry point for checksums calculation component.
+/// Should be called from "main".
+/// Accepts notifications about epoch change, or changes in specific bubblegum tree or account,
+/// and schedules checksum calculation.
+pub fn run_bg_consistency_calculator(
+    mut rcv: Receiver<ConsistencyCalcMsg>,
+    storage: Arc<Storage>,
+    force_reingestable_slot_getter: Arc<impl SlotGetter + Send + Sync + 'static>,
+    mut shutdown_signal: tokio::sync::broadcast::Receiver<()>,
+    metrics: Arc<ChecksumCalculationMetricsConfig>,
+) {
+    tokio::spawn(async move {
+        let bbgm_tasks: Arc<Mutex<BTreeSet<BbgmTask>>> = Arc::new(Mutex::new(BTreeSet::new()));
+        let acc_tasks: Arc<Mutex<BTreeSet<AccTask>>> = Arc::new(Mutex::new(BTreeSet::new()));
+
+        // Taks that calculates bubblegum checksums
+        let _bbgm_bg = {
+            let storage = storage.clone();
+            let bbgm_tasks = bbgm_tasks.clone();
+            let force_reingestable_slot_getter = force_reingestable_slot_getter.clone();
+            let metrics = metrics.clone();
+            tokio::spawn(async move {
+                process_bbgm_tasks(storage, bbgm_tasks, force_reingestable_slot_getter, metrics)
+                    .await;
+            })
+        };
+        // Taks that calculates account NFT checksums
+        let _acc_bg = {
+            let storage = storage.clone();
+            let acc_tasks = acc_tasks.clone();
+            let metrics = metrics.clone();
+            tokio::spawn(async move {
+                process_acc_tasks(storage, acc_tasks, metrics).await;
+            })
+        };
+
+        loop {
+            let calc_msg = tokio::select! {
+                msg = rcv.recv() => msg,
+                _ = shutdown_signal.recv() => {
```

Comment:
nit: as far as this runs inside simple tokio task that is not related to any JoinSet, shutdown_signal is not really needed here because we do not wait for this task to complete anywhere